### ISSUE TITLE
C++   set_allow_soft_placement does not work

### ISSUE THREAD

**Issue by xu-song** [tensorflow/tensorflow#22304]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
- **OS Platform and Distribution: Mac  CPU
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.8.0
- **Python version**: 2.7

### Describe the problem

I have set set_allow_soft_placement(true), but it does not work.

### Source code / logs

c++ source code
```c++
    const string pathToGraph = “tf/checkpoint/model-1.meta";

    tensorflow::SessionOptions options;
    options.config.set_allow_soft_placement(true);  // why this does not work?
    auto session = NewSession(options);

    Status status = session->Create(graph_def.graph_def());

```

It get errors in my mac(cpu),  no error in GPU.

```sh
./eval
libc++abi.dylib: terminating with uncaught exception of type std::runtime_error: 
Error creating graph: Invalid argument: Cannot assign a device for 
operation 'ExponentialMovingAverage/AssignMovingAvg_5/sub': 
Operation was explicitly assigned to /device:GPU:0 but available devices are 
[ /job:localhost/replica:0/task:0/device:CPU:0 ]. 
Make sure the device specification refers to a valid device.
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by harshini-gadige** [comment]
@xu-song Hi, can you share an example to reproduce these errors ?

**Comment by xu-song** [comment]
ok, i will prepare one example soon

**Comment by xu-song** [comment]
# simple example

this is a simple example, which get `result = a*b+1`

## python - save model
```py
import numpy as np
import tensorflow as tf
import sys, os

if __name__ == '__main__':
    train_dir = os.path.join('demo_model/', "demo")

    with tf.device('/gpu:0'):
        a = tf.placeholder(dtype=tf.int32, shape=None, name='a')
        b = tf.placeholder(dtype=tf.int32, shape=None, name='b')
        y = tf.Variable(tf.ones(shape=[1], dtype=tf.int32), dtype=tf.int32, name='y')
        res = tf.add(tf.multiply(a, b), y, name='res')

    config = tf.ConfigProto(allow_soft_placement=True)
    with tf.Session(config=config) as sess:
        feed_dict = dict()
        feed_dict[a] = 2
        feed_dict[b] = 3
        fetch_list = [res]
        sess.run(tf.global_variables_initializer())
        saver = tf.train.Saver()
        
        res = sess.run(feed_dict=feed_dict, fetches=fetch_list)
        saver.save(sess, train_dir)

        print("result: ", res[0])
```

## C++ restore model
```
#include <iostream>
#include "tensorflow/core/public/session.h"
#include "tensorflow/core/protobuf/meta_graph.pb.h"
#include "tensorflow/cc/client/client_session.h"
#include "tensorflow/cc/ops/standard_ops.h"
#include "tensorflow/core/framework/tensor.h"

using namespace std;
using namespace tensorflow;

int main()
{
    const string pathToGraph = "demo_model/demo.meta";
    const string checkpointPath = "demo_model/demo";
    SessionOptions options;
    options.config.set_allow_soft_placement(true);   // this line does not work
    auto session = NewSession(options); 

//    auto session = NewSession(SessionOptions());
    if (session == nullptr)
    {
        throw runtime_error("Could not create Tensorflow session.");
    }

    Status status;
    MetaGraphDef graph_def;
    status = ReadBinaryProto(Env::Default(), pathToGraph, &graph_def);
    if (!status.ok())
    {
        throw runtime_error("Error reading graph definition from " + pathToGraph + ": " + status.ToString());
    }

// 利用读入的模型的图的拓扑结构构建一个session
    status = session->Create(graph_def.graph_def());
    if (!status.ok())
    {
        throw runtime_error("Error creating graph: " + status.ToString());
    }

// 读入预先训练好的模型的权重
    Tensor checkpointPathTensor(DT_STRING, TensorShape());
    checkpointPathTensor.scalar<std::string>()() = checkpointPath;
    status = session->Run(
            {{ graph_def.saver_def().filename_tensor_name(), checkpointPathTensor },},
            {},
            {graph_def.saver_def().restore_op_name()},
            nullptr);
    if (!status.ok())
    {
        throw runtime_error("Error loading checkpoint from " + checkpointPath + ": " + status.ToString());
    }

    std::vector<std::pair<string, Tensor>> input;
    Tensor a(tensorflow::DT_INT32, tensorflow::TensorShape({1}));
    Tensor b(tensorflow::DT_INT32, tensorflow::TensorShape({1}));

    auto a_map = a.tensor<int,1>();
    a_map(0) = 2;
    auto b_map = b.tensor<int,1>();
    b_map(0) = 3;
    input.emplace_back(std::string("a"), a);
    input.emplace_back(std::string("b"), b);

//   运行模型，并获取输出
    std::vector<tensorflow::Tensor> answer;
    status = session->Run(input, {"res"}, {}, &answer);

    Tensor result = answer[0];
    auto result_map = result.tensor<int,1>();
    cout<<"result: "<<result_map(0)<<endl;

    return 0;
}
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @reedwm: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @reedwm: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @reedwm: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @reedwm: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by hahadashi** [comment]
> # simple example
> this is a simple example, which get `result = a*b+1`
> 
> ## python - save model
> ```python
> # demo.py
> import tensorflow as tf
> import os
> 
> if __name__ == '__main__':
>     train_dir = os.path.join('demo_model/', "demo")
> 
>     with tf.device('/gpu:0'):
>         a = tf.placeholder(dtype=tf.int32, shape=None, name='a')
>         b = tf.placeholder(dtype=tf.int32, shape=None, name='b')
>         y = tf.Variable(tf.ones(shape=[1], dtype=tf.int32), dtype=tf.int32, name='y')
>         res = tf.add(tf.multiply(a, b), y, name='res')
> 
>     config = tf.ConfigProto(allow_soft_placement=True)
>     with tf.Session(config=config) as sess:
>         feed_dict = {a:2, b:3}
>         fetch_list = [res]
>         sess.run(tf.global_variables_initializer())
>         saver = tf.train.Saver()
> 
>         res = sess.run(feed_dict=feed_dict, fetches=fetch_list)
>         saver.save(sess, train_dir)
> 
>         print("result: ", res[0])
> ```
> 
> ```shell
> $ python demo.py 
> 2018-09-29 09:35:28.855859: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
> ('result: ', array([7], dtype=int32))
> ```
> 
> ## python - restore
> ```python
> # demo-restore.py
> import tensorflow as tf
> import os
> import numpy as np
> 
> if __name__ == '__main__':
>     config = tf.ConfigProto(allow_soft_placement=True)
>     with tf.Session(config=config) as sess:
>         saver = tf.train.import_meta_graph('demo_model/demo.meta')
>         saver.restore(sess, tf.train.latest_checkpoint('demo_model/'))
>         graph = tf.get_default_graph()
>         a = graph.get_tensor_by_name("a:0")
>         b = graph.get_tensor_by_name("b:0")
>         feed_dict = {a: 2, b: 3}
> 
>         op_to_restore = graph.get_tensor_by_name("res:0")
>         print(sess.run(fetches=op_to_restore, feed_dict=feed_dict))
> ```
> 
> ```
> $ python demo-restore.py
> [7]
> ```
> 
> ## C++ restore model
> ```c
> // demo.cc
> #include <iostream>
> #include "tensorflow/core/public/session.h"
> #include "tensorflow/core/protobuf/meta_graph.pb.h"
> #include "tensorflow/cc/client/client_session.h"
> #include "tensorflow/cc/ops/standard_ops.h"
> #include "tensorflow/core/framework/tensor.h"
> 
> using namespace std;
> using namespace tensorflow;
> 
> int main()
> {
>     const string pathToGraph = "demo_model/demo.meta";
>     const string checkpointPath = "demo_model/demo";
>     SessionOptions options;
>     options.config.set_allow_soft_placement(true);   // this line does not work
>     auto session = NewSession(options); 
> 
> //    auto session = NewSession(SessionOptions());
>     if (session == nullptr)
>     {
>         throw runtime_error("Could not create Tensorflow session.");
>     }
> 
>     Status status;
>     MetaGraphDef graph_def;
>     status = ReadBinaryProto(Env::Default(), pathToGraph, &graph_def);
>     if (!status.ok())
>     {
>         throw runtime_error("Error reading graph definition from " + pathToGraph + ": " + status.ToString());
>     }
> 
>    // restore graph
>     status = session->Create(graph_def.graph_def());
>     if (!status.ok())
>     {
>         throw runtime_error("Error creating graph: " + status.ToString());
>     }
> 
>     //  restore checkpoint 
>     Tensor checkpointPathTensor(DT_STRING, TensorShape());
>     checkpointPathTensor.scalar<std::string>()() = checkpointPath;
>     status = session->Run(
>             {{ graph_def.saver_def().filename_tensor_name(), checkpointPathTensor },},
>             {},
>             {graph_def.saver_def().restore_op_name()},
>             nullptr);
>     if (!status.ok())
>     {
>         throw runtime_error("Error loading checkpoint from " + checkpointPath + ": " + status.ToString());
>     }
> 
>     std::vector<std::pair<string, Tensor>> input;
>     Tensor a(tensorflow::DT_INT32, tensorflow::TensorShape({1}));
>     Tensor b(tensorflow::DT_INT32, tensorflow::TensorShape({1}));
> 
>     auto a_map = a.tensor<int,1>();
>     a_map(0) = 2;
>     auto b_map = b.tensor<int,1>();
>     b_map(0) = 3;
>     input.emplace_back(std::string("a"), a);
>     input.emplace_back(std::string("b"), b);
> 
>    //   run session
>     std::vector<tensorflow::Tensor> answer;
>     status = session->Run(input, {"res"}, {}, &answer);
> 
>     Tensor result = answer[0];
>     auto result_map = result.tensor<int,1>();
>     cout<<"result: "<<result_map(0)<<endl;
> 
>     return 0;
> }
> ```
> 
> ## error - in cpu
> ```
> ./demo
> libc++abi.dylib: terminating with uncaught exception of type std::runtime_error: Error creating graph: 
> Invalid argument: Cannot assign a device for operation 'y': Operation was explicitly assigned to 
> /device:GPU:0 but available devices are [ /job:localhost/replica:0/task:0/device:CPU:0 ]. Make sure the 
> device specification refers to a valid device.
> 	 [[Node: y = VariableV2[_output_shapes=[[1]], container="", dtype=DT_INT32, shape=[1], 
> shared_name="", _device="/device:GPU:0"]()]]
> Abort trap: 6
> ```
> 
> ## no error - in gpu
because you train your model with gpu,you can try load your model just cpu type

**Comment by chunduriv** [comment]
@xu-song,

Sorry for late response. Can you please a take a look at this [comment](https://github.com/tensorflow/tensorflow/issues/22304#issuecomment-555813956) ?

Whether soft placement is allowed. If allow_soft_placement is true,
 an op will be placed on CPU if
   1. there's no GPU implementation for the OP
 or
   2. no GPU devices are known or registered
 or
   3. need to co-locate with reftype input(s) which are from CPU.

Thanks!

**Comment by formath** [comment]
> 

Does this problem have been fixed?
